# <span style="color:teal;">CIS 211 Project 7:  SQLite</span>

##### Due 11:00 P.M. Friday Feb 26

This project will give you some experience writing database queries in SQL and connecting to a SQLite database from Python.

### <span style="color:teal">Setup</span>

Download the Sakila database (the file named `sakila211.db`) and move it to the same folder as this IPython notebook.

### <span style="color:teal">Executing Queries in an IPython Notebook</span>

To solve the problems in the first two sections of this project you need to write a SQL query based on the tables in the Sakila211 database.  

To execute a query from an IPython Notebook write the query in quotes as part of a shell command:
```
! sqlite3 sakila211.db "..."
```

For example, to answer the question "What are the names and ratings of the films" your code cell would have
```
! sqlite3 sakila211.db "select title, rating from film"
```

Some of the queries will produce a lot of output.  You can add a `LIMIT` clause to any query if you want to reduce the size of the output.

**Note:** &nbsp; Your grade will not depend on the output format.  If want to have a nicer looking output that includes column headers create a file named `.sqliterc` in your home folder and add these two lines:
```
.header on
.mode column
```

## <span style="color:teal">Part 1:  &nbsp; Queries</span>

The following questions can all be answered using records from a single table.

(1) What are the first and last names of all the customers?

In [144]:
! sqlite3 sakila211.db "select first_name, last_name from customer"

Mary|Smith
Patricia|Johnson
Linda|Williams
Barbara|Jones
Elizabeth|Brown
Jennifer|Davis
Maria|Miller
Susan|Wilson
Margaret|Moore
Dorothy|Taylor
Lisa|Anderson
Nancy|Thomas
Karen|Jackson
Betty|White
Helen|Harris
Sandra|Martin
Donna|Thompson
Carol|Garcia
Ruth|Martinez
Sharon|Robinson
Michelle|Clark
Laura|Rodriguez
Sarah|Lewis
Kimberly|Lee
Deborah|Walker
Jessica|Hall
Shirley|Allen
Cynthia|Young
Angela|Hernandez
Melissa|King
Brenda|Wright
Amy|Lopez
Anna|Hill
Rebecca|Scott
Virginia|Green
Kathleen|Adams
Pamela|Baker
Martha|Gonzalez
Debra|Nelson
Amanda|Carter
Stephanie|Mitchell
Carolyn|Perez
Christine|Roberts
Marie|Turner
Janet|Phillips
Catherine|Campbell
Frances|Parker
Ann|Evans
Joyce|Edwards
Diane|Collins
Alice|Stewart
Julie|Sanchez
Heather|Morris
Teresa|Rogers
Doris|Reed
Gloria|Cook
Evelyn|Morgan
Jean|Bell
Cheryl|Murphy
Mildred|Bailey
Katherine|Rivera
Joan|Cooper
Ashley|Richardson
Judith|Cox
Rose|Howard
Janice|Ward
Kelly|Torres
Nicole|Peterson
Judy|Gray
Christina|Ramirez
Kathy|James
Theresa

(2) What are the e-mail addresses of customers whose first name is “Willard”?

In [145]:
! sqlite3 sakila211.db "select first_name, email from customer where first_name='Willard'"

Willard|willard.lumpkin@sakilacustomer.org


(3) How many customers are there?

In [146]:
! sqlite3 sakila211.db "select count() from customer"

599


(4) How many customers shop at store number 1?

In [147]:
! sqlite3 sakila211.db "select count() from customer where store_id=1"

326


(5) How much does it cost to rent the film named “Virtual Spoilers”?

In [148]:
! sqlite3 sakila211.db "select rental_rate from film where title='Virtual Spoilers'"

4.99


(6) Do any films have the word “Princess” in the title?

In [149]:
! sqlite3 sakila211.db "select title from film where title like '%Princess%' "

Image Princess
Princess Giant


(7) What are the titles of the films that are longer than 180 minutes?

In [150]:
! sqlite3 sakila211.db "select title, length from film where length>80 "

Academy Dinosaur|86
Affair Prejudice|117
African Egg|130
Agent Truman|169
Alabama Devil|114
Alamo Videotape|126
Alaska Phantom|136
Ali Forever|150
Alice Fantasia|94
Alley Evolution|180
Alone Trip|82
Amadeus Holy|113
American Circus|129
Amistad Midsummer|85
Anaconda Confessions|92
Analyze Hoosiers|181
Annie Identity|86
Anonymous Human|179
Anthem Luke|91
Antitrust Tomatoes|168
Anything Savannah|82
Apache Divine|92
Apocalypse Flamingos|119
Apollo Teen|153
Arachnophobia Rollercoaster|147
Argonauts Town|127
Arizona Bang|121
Armageddon Lost|99
Army Flintstones|148
Arsenic Independence|137
Artist Coldblooded|170
Atlantis Cause|170
Attacks Hate|113
Attraction Newton|83
Autumn Crow|108
Baby Hall|153
Backlash Undefeated|118
Badman Dawn|162
Baked Cleopatra|182
Ballroom Mockingbird|173
Bang Kwai|87
Banger Pinocchio|113
Barefoot Manchurian|129
Basic Easy|90
Beach Heartbreakers|122
Bear Graceland|160
Beast Hunchback|89
Beauty Grease|175
Bed Highball|106
Beethoven Exorcist|151
Behavior Runaway|100
Be

(8) How many films have a “G” rating and are less than 60 minutes long?

In [151]:
! sqlite3 sakila211.db "select count() from film where rating='G' and length<60"

21


(9) What is the maximum replacement cost for any film?

In [152]:
! sqlite3 sakila211.db "select max(replacement_cost) from film"

29.99


(10) Print a table that lists the different types of ratings and the number of films that have that rating.

In [153]:
! sqlite3 sakila211.db "select rating, count() from film  group by rating"

G|178
NC-17|210
PG|194
PG-13|223
R|195


## <span style="color:teal">Part 2: &nbsp; Joins</span>

The following queries all require a join of two or more tables. As a hint for how to create the query the table names are shown next to each a question.

(11) What are the film IDs of movies starring Jude Cruise?  &nbsp; &nbsp; [`actor`, `film_actor`]

In [154]:
! sqlite3 sakila211.db "select film_id from actor join film_actor using (actor_id) where actor.first_name='Jude' and actor.last_name='Cruise'"

16
34
101
114
122
134
144
153
192
213
258
267
317
340
393
437
447
502
592
605
637
685
707
714
717
737
767
852
891
918


(12) What language was “Ace Goldfinger” filmed in?  &nbsp; &nbsp; [`film`, `language`]

In [155]:
! sqlite3 sakila211.db "select name from film join language using (language_id) where film.title='Ace Goldfinger'"

Japanese


(13) How many actors starred in the film named “Scalawag Duck”? &nbsp; &nbsp; [`film`, `film_actor`]

In [156]:
! sqlite3 sakila211.db "select count() from film join film_actor using (film_id) where film.title= 'Scalawag Duck'"

5


(14) Produce a table that shows how many films each customer has rented.  &nbsp; &nbsp; [`customer`, `rental`]

In [157]:
! sqlite3 sakila211.db "select first_name, last_name, count() from customer join rental using (customer_id) group by customer_id"

Mary|Smith|32
Patricia|Johnson|27
Linda|Williams|26
Barbara|Jones|22
Elizabeth|Brown|38
Jennifer|Davis|28
Maria|Miller|33
Susan|Wilson|24
Margaret|Moore|23
Dorothy|Taylor|25
Lisa|Anderson|24
Nancy|Thomas|28
Karen|Jackson|27
Betty|White|28
Helen|Harris|32
Sandra|Martin|28
Donna|Thompson|21
Carol|Garcia|22
Ruth|Martinez|24
Sharon|Robinson|30
Michelle|Clark|35
Laura|Rodriguez|22
Sarah|Lewis|30
Kimberly|Lee|25
Deborah|Walker|29
Jessica|Hall|34
Shirley|Allen|31
Cynthia|Young|32
Angela|Hernandez|36
Melissa|King|34
Brenda|Wright|26
Amy|Lopez|29
Anna|Hill|21
Rebecca|Scott|24
Virginia|Green|32
Kathleen|Adams|27
Pamela|Baker|23
Martha|Gonzalez|34
Debra|Nelson|29
Amanda|Carter|27
Stephanie|Mitchell|25
Carolyn|Perez|30
Christine|Roberts|24
Marie|Turner|26
Janet|Phillips|27
Catherine|Campbell|34
Frances|Parker|22
Ann|Evans|17
Joyce|Edwards|28
Diane|Collins|35
Alice|Stewart|33
Julie|Sanchez|29
Heather|Morris|30
Teresa|Rogers|29
Doris|Reed|22
Gloria|Cook|30
Evelyn|Morgan|28
Jean|Bell|27
Cheryl|Murphy

(15) How many films did the customer named Smith rent?  &nbsp; &nbsp; [`customer`, `rental`]

In [158]:
! sqlite3 sakila211.db "select count() from customer join rental using (customer_id) where first_name='Smith' or last_name='Smith'

32


(16) Which customers have not returned films (*i.e.* the return_date field in the rental table is null)?  &nbsp; &nbsp; [`customer`, `rental`]

In [159]:
! sqlite3 sakila211.db "select first_name, last_name from customer join rental using (customer_id) where return_date is NULL"

Gail|Knight
Gregory|Mauldin
Louise|Jenkins
Willie|Howell
Emily|Diaz
Laurie|Lawrence
Lisa|Anderson
Freddie|Duggan
Heather|Morris
Roland|South
Natalie|Meyer
Scott|Shelley
Louis|Leone
Cathy|Spencer
Guy|Brownlee
Justin|Ngo
Jerry|Jordon
Alicia|Mills
Jeanne|Lawson
Dustin|Gillette
Morris|Mccarter
Judy|Gray
Julie|Sanchez
Ramona|Hale
Sonia|Gregory
Margie|Wade
Zachary|Hite
Jordan|Archuleta
Andy|Vanhorn
Elmer|Noe
Peggy|Myers
Sergio|Stanfield
Laura|Rodriguez
Anna|Hill
Gail|Knight
Marilyn|Ross
Mildred|Bailey
Grace|Ellis
Barry|Lovelace
Edward|Baugh
Cassandra|Walters
Bill|Gavin
Curtis|Irby
Allison|Stanley
Courtney|Day
Gilbert|Sledge
Ivan|Cromwell
Wanda|Patterson
Viola|Hanson
Stacey|Montgomery
Justin|Ngo
Cecil|Vines
Leonard|Schofield
Craig|Morrell
Marion|Snyder
Stacy|Cunningham
Lois|Butler
Elmer|Noe
Cynthia|Young
Sonia|Gregory
Natalie|Meyer
Marcia|Dean
Ana|Bradley
Joshua|Mark
Cassandra|Walters
Wendy|Harrison
Florence|Woods
Bernard|Colby
Beverly|Brooks
Annette|Olson
Chris|Brothers
Tyler|Wren
Elizabeth|

(17) Which actors starred in “Splash Gump”?  &nbsp; &nbsp; [`film`, `film_actor`, `actor`]

In [160]:
! sqlite3 sakila211.db "select first_name, last_name from actor join film_actor using(actor_id) join film using (film_id) where film.title='Splash Gump'"

Penelope|Guiness
Jennifer|Davis
Uma|Wood
Cameron|Streep
Christian|Neeson
Ray|Johansson
Gene|Willis
Morgan|Williams
Humphrey|Willis
Al|Garland


(18) Which films has Penelope Guiness starred in?  &nbsp; &nbsp; [`film`, `film_actor`, `actor`]

In [161]:
! sqlite3 sakila211.db "select title from actor join film_actor using(actor_id) join film using (film_id) where actor.first_name='Penelope' and actor.last_name='Guiness'"

Academy Dinosaur
Anaconda Confessions
Angels Life
Bulworth Commandments
Cheaper Clyde
Color Philadelphia
Elephant Trojan
Gleaming Jawbreaker
Human Graffiti
King Evolution
Lady Stage
Language Cowboy
Mulholland Beast
Oklahoma Jumanji
Rules Human
Splash Gump
Vertigo Northwest
Westward Seabiscuit
Wizard Coldblooded


(19) What are the names of the films rented by the customer named Knight?  &nbsp; &nbsp; [`customer`, `rental`, `inventory`, `film`]

In [162]:
! sqlite3 sakila211.db "select title from customer join rental using (customer_id) join inventory using (inventory_id) join film using (film_id) where last_name='Knight'"

Jericho Mulan
Frost Head
Clerks Angels
Wife Turn
Expecations Natural
Lawless Vision
Soup Wisdom
Heartbreakers Bright
Bikini Borrowers
Attacks Hate
Cupboard Sinners
Princess Giant
Cyclone Family
Platoon Instinct
Knock Warlock
Excitement Eve
Right Cranes
Hyde Doctor
Fatal Haunted
Chasing Fight
Falcon Volume
Tomorrow Hustler
Half Outfield
Horror Reign
Borrowers Bedazzled


## <span style="color:teal">Part 3: &nbsp; Monthly Report</span>

For this part of the project you will write a report generator that will print a summary of monthly activity for a customer.  The report should show a list of all the films rented during the month along with a total amount due.
This example shows the activity for the customer named Leo Ebert for the month of June, 2005.

<pre>
--- Sakila DVD Rentals ---

Monthly Report for Leo Ebert 

Lonely Elephant                 06/16/2005    $2.99
                   ** late fee  06/24/2005    $2.99
Enemy Odds                      06/18/2005    $4.99
                   ** late fee  06/27/2005    $4.99
Song Hedwig                     06/20/2005    $0.99
                   ** late fee  06/26/2005    $0.99
Wash Heavenly                   06/21/2005    $4.99
Hurricane Affair                06/21/2005    $2.99

Monthly Total:  $25.92
</pre>

To compute the total, the report generator needs to add up all the rental fees, and for any film returned late, add an additional rental fee as the late charge. In the example, Leo rented *Lonely Elephant* on June 16 but didn’t return it until June 24. It was a 3-day rental, so he was charged an additional \$2.99 (it was 5 days late, so we could have charged him \$4.98 in late fees, but we’ll keep it simple and just charge the basic fee as the late fee).

We've broken the problem into separate parts so they can be implemented and tested separately.  Your job is to fill in the code cells so each part passes the auto-grader tests.

**Note:** &nbsp; There is no documentation requirement for this project.

### <span style="color:teal">Database Connection</span>

To simplify the code for this program you can define the database connection as a global variable.  The following code cell imports the libraries you need and defines the database connection.

In [163]:
from sqlite3 import connect

db = connect('sakila211.db')

### <span style="color:teal">Dates</span>

Dates in a SQLite database are stored as strings.  The function below uses a Python library named `datetime` to parse a date string and return a string that can be printed as part of the report.  Read the online documentation for the `datetime` module if you want to learn more about date formats and functions that parse and format dates.

**Note:** &nbsp; The format string is a global variable -- you can use this same format in your function that checks to see if a CD is overdue.

In [164]:
from datetime import datetime

date_format = "%Y-%m-%d %H:%M:%S.%f"

def mmddyyyy(date):
    d = datetime.strptime(date, date_format)
    return '{:02d}/{:02d}/{:02d}'.format(d.month, d.day, d.year)

In [165]:
assert mmddyyyy('2005-02-26 20:59:35.000') == '02/26/2005'

### <span style="color:teal">Customer Name</span>

Fill in the definition of the `customer_name` function.  It will be passed a string with a last name, and it should return a list of tuples of the form `(first, last)` with the first and last names of all customers with the same last name.  If there are no customers with that name the list should be empty.

In [166]:
def customer_names(name):
    name=name.title()
    t=(name,)
    names=db.execute('select first_name, last_name from customer where last_name=?',  t)
    name_list=[]
    for row in names:
        name_list.append(row)
    return name_list

In [167]:
customer_names('turing')

[]

##### <span style="color:red">Autograder Tests:</span>

**Do not edit, copy, or delete these cells.**

In [168]:
assert customer_names('ebert') == [('Leo', 'Ebert')]

In [169]:
assert customer_names('smith') == [('Mary', 'Smith')]

In [170]:
assert customer_names('turing') == []

### <span style="color:teal">Film List</span>

Fill in the body of the `film_list` function so it returns a list of all CDs rented by a customer during a specified month. 
The parameters that will be passed to the function are the customer's last name (a string) and the month and year (both integers).  Return a list of tuples that have the following values, in the order shown:
* film title
* rental rate (how much it costs to rent the CD)
* rental date (the date the CD was rented)
* return date (the date the CD was returned)
* rental duration (the number of days the CD can be out before a late fee is charged) 

All of the information you need can be retrieved with a single query that joins the rental, inventory, film, and customer tables. Note that this is the same query used to answer Question 19 above, except with different column names.

**Hint:** &nbsp; The dates in the database are strings.  Think about pattern matching when you write the WHERE clause that specifies the date range.


In [171]:
def film_list(name, month, year):
    if len(str(month))<2:
        month=str("0"+str(month))
    
    final_list=[]
    
    ## here's the pattern mathching
    d=("%{}-{}%".format(year,month),)
    
    #I could not get the database to narrow down the results for both
    #the time period and the last name I kept getting a multitude
    #of errors. I had to narrow time first
    #and write a loop to find the right people
    #below is my failed attemps to do it all in one step
    # Note: n=(name,)
    
    #work_list=db.execute ("select count() from customer join rental using(customer_id)join inventory using (inventory_id) join film where last_name=? and rental_date like ?",(n,d))
    
    
    
    
    work_list=db.execute ("select last_name, title, rental_rate, rental_date, return_date,rental_duration  from customer join rental using(customer_id)join inventory using (inventory_id) join film using (film_id) where rental_date like ?",(d))
    
    for item in work_list:
        if item[0]==name:
            final_list.append(item[1:])
    return final_list
    

##### <span style="color:red">Autograder Tests:</span>

**Do not edit, copy, or delete these cells.**

In [172]:
leo = film_list('Ebert', 6, 2005)
assert len(leo) == 5
assert leo[0] == (('Lonely Elephant', 2.99, '2005-06-16 20:59:35.000', '2005-06-24 15:31:35.000', 3))

In [173]:
mary = film_list('Smith', 8, 2005)
assert len(mary) == 11
assert mary[0] == ('Adaptation Holes', 2.99, '2005-08-01 08:51:04.000', '2005-08-10 12:12:04.000', 7)

### <span style="color:teal">Overdue CDs</span>

To decide whether a CD was returned late you need to subtract the rental date from the return date and see if the difference is greater than the rental duration. The easiest way to do this is to use Python’s `datetime` library.

Using a call to `strptime` (see the body of the `mmddyyyy` function defined above) create `datetime` objects for the rental date and the return date. Then subtract the rental date from the return date to get a `timedelta` object:
```
    diff = returned - rented
```
Now you can get attributes of the object named `diff` to tell you how far apart the dates were, for example `diff.days` will tell you the time difference in days.

Fill in the body of the `overdue` function.  The parameters are a rental date, a return date, and the rental duration.  Return TRUE if the CD was not returned or if the number of days it was checked out is greater than the rental duration.

In [174]:
from datetime import datetime
def overdue(rent_date, return_date, duration):
    date_format = "%Y-%m-%d %H:%M:%S.%f"

    rented= datetime.strptime(rent_date, date_format)
    returned=datetime.strptime(return_date, date_format)
    diff=returned-rented
    diff=diff.days
    if (diff>duration):
        return True
    


In [175]:
assert overdue('2005-06-16 20:59:35.000', '2005-06-19 20:59:35.000', 2)

In [176]:
assert not overdue('2005-06-16 20:59:35.000', '2005-06-17 20:59:35.000', 2)

In [177]:
assert not overdue('2005-06-16 20:59:35.000', '2005-06-19 20:59:34.000', 2)

### <span style="color:teal">Report Generator</span>

Fill in the body of the function named `sakila_report` so it prints a monthly report.  The arguments passed to the function will be a customer's last name (a string) and a month and year (both integers).

If the customer is not in the database, or if there are two customers with the same last name, return None.  Otherwise print a list of films and late fees and the total amount due.

You can format the report however you would like, as long as the following items appear somewhere in the output:
* every fee should be printed with a dollar sign and an amount, for example `$2.99` or `25.92`
* the total amound due should be printed on a line that includes the word `"Total"` followed by a dollar amount on the same line

In [187]:
def sakila_report(customer, month, year):

    
    cust=customer_names(customer)
    Total=float(0)
   
   
    if cust!=[]:
        print("--- Sakila DVD Rentals ---")
        first_name,last_name=cust[0]
        print("Monthly Report For "+first_name+" "+last_name)
    
        data=film_list(last_name, month, year)
        for movie in data:
            print(str(movie[0])+"                   " +str(movie [2])+"    $"+str(movie[1]))
             
            Total+=movie[1]
            if overdue(movie[2], movie[3], movie[4]):
            
                print("                         ***Late Fee    "+ movie[3]+"   $"+str(movie[1]))
                Total+=movie[1]
        print("Monthly Total=   $"+str("{0:.2f}".format(Total)))
  

##### <span style="color:red">Tests</span>

You can use code cells in this section (and add more if you wish) to test your functions.

In [188]:
sakila_report('Smith', 8, 2005)

--- Sakila DVD Rentals ---
Monthly Report For Mary Smith
Adaptation Holes                   2005-08-01 08:51:04.000    $2.99
                         ***Late Fee    2005-08-10 12:12:04.000   $2.99
Racer Egg                   2005-08-02 15:36:52.000    $2.99
                         ***Late Fee    2005-08-10 16:40:52.000   $2.99
Finding Anaconda                   2005-08-02 18:01:38.000    $0.99
Minds Truman                   2005-08-17 12:37:54.000    $4.99
Dalmations Sweden                   2005-08-18 03:57:29.000    $0.99
Patient Sister                   2005-08-19 09:55:16.000    $0.99
Jumanji Blade                   2005-08-19 13:56:54.000    $2.99
Unforgiven Zoolander                   2005-08-21 23:33:57.000    $0.99
Fireball Philadelphia                   2005-08-22 01:27:57.000    $0.99
                         ***Late Fee    2005-08-27 07:01:57.000   $0.99
Fireball Philadelphia                   2005-08-22 19:41:37.000    $0.99
                         ***Late Fee    2005-08-

##### <span style="color:red">Autograder Tests</span>

The tests for the `sakila_report` function will call the function and while using the IPython `capture` command to save everything the function prints.  We'll then use a function that counts the number of times a dollar amount (from either a rental fee or a late fee) occurs and make sure the total amount is the expected value.

**Do not edit, copy, or delete these cells.**

In [180]:
import re

def check_report(text, expected_count, expected_total):
    nums = []
    for line in text.split('\n'):
        has_total = re.search(r'Total.*(\$\d+\.\d\d)', line)
        if has_total:
            assert has_total.group(1) == expected_total
        else:
            for s in re.findall(r'\$\d+\.\d\d', line):
                nums.append(s)
    assert(len(nums) == expected_count)
    total = 0.0
    for s in nums:
        total += float(s[1:])
    assert(round(total,2)) == round(float(expected_total[1:]), 2)
    return True

In [181]:
%%capture res
sakila_report('Ebert', 6, 2005)

In [182]:
assert check_report(res.stdout, 8, '$25.92')

In [183]:
%%capture res
sakila_report('Smith', 8, 2005)

In [184]:
assert check_report(res.stdout, 15, '$32.85')

In [185]:
%%capture res
sakila_report('Turing', 6, 2005)

In [186]:
assert len(res.stdout) == 0